src/System.Security.Cryptography.DeriveBytes/src/System/Security/Cryptography/Rfc2898DeriveBytes.cs

```diff
@@ -0,0 +1,208 @@
+                Initialize();
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (_hmacSha1 != null)
+                    _hmacSha1.Dispose();
+                _hmacSha1 = null;
+                if (_buffer != null)
+                    Array.Clear(_buffer, 0, _buffer.Length);
+                if (_password != null)
+                    Array.Clear(_password, 0, _password.Length);
+                if (_salt != null)
+                    Array.Clear(_salt, 0, _salt.Length);
+            }
+            base.Dispose(disposing);
+        }
+
+        public override byte[] GetBytes(int cb)
+        {
+            if (cb <= 0)
+                throw new ArgumentOutOfRangeException("cb", SR.ArgumentOutOfRange_NeedPosNum);
+            byte[] password = new byte[cb];
+
+            int offset = 0;
+            int size = _endIndex - _startIndex;
+            if (size > 0)
+            {
+                if (cb >= size)
+                {
+                    Array.Copy(_buffer, _startIndex, password, 0, size);
+                    _startIndex = _endIndex = 0;
+                    offset += size;
+                }
+                else
+                {
+                    Array.Copy(_buffer, _startIndex, password, 0, cb);
+                    _startIndex += cb;
+                    return password;
+                }
+            }
+
+            Debug.Assert(_startIndex == 0 && _endIndex == 0, "Invalid start or end index in the internal buffer.");
+
+            while (offset < cb)
+            {
+                byte[] T_block = Func();
+                int remainder = cb - offset;
+                if (remainder > BlockSize)
+                {
+                    Array.Copy(T_block, 0, password, offset, BlockSize);
+                    offset += BlockSize;
+                }
+                else
+                {
+                    Array.Copy(T_block, 0, password, offset, remainder);
+                    offset += remainder;
+                    Array.Copy(T_block, remainder, _buffer, _startIndex, BlockSize - remainder);
+                    _endIndex += (BlockSize - remainder);
+                    return password;
+                }
+            }
+            return password;
+        }
+
+        public override void Reset()
+        {
+            Initialize();
+        }
+
+        private void Initialize()
+        {
+            if (_buffer != null)
+                Array.Clear(_buffer, 0, _buffer.Length);
+            _buffer = new byte[BlockSize];
+            _block = 1;
+            _startIndex = _endIndex = 0;
+        }
+
+        // This function is defined as follows:
+        // Func (S, i) = HMAC(S || i) | HMAC2(S || i) | ... | HMAC(iterations) (S || i) 
+        // where i is the block number.
+        private byte[] Func()
+        {
+            byte[] INT_block = Helpers.Int(_block);
```

Comment:
Rather than having Int allocate a new byte[], how about having it accept a byte[] and offset to write into and at?  Then you can avoid the extra allocation and copy by just having it generate the values directly into temp below.